src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h

```diff
@@ -0,0 +1,198 @@
+            else if (req->kind == MPIR_REQUEST_KIND__RECV)         \
+                MPIDIG_gpu_unstage_recv_buf(gpu_stage_buf_);       \
+            MPIDIG_GPU_REQUEST(req, stage_buf) = NULL;             \
+        }                                                          \
+    } while (0)
+
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_get_buf(size_t size)
+{
+    void *buf = NULL;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_GET_BUF);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_GET_BUF);
+    MPIDIU_buf_pool_t **pools = MPIDIG_gpu_global.buf_pool;
+    for (int i = 0; i < MPIDIG_GPU_BUF_POOL_NUM; i++) {
+        if (size > pools[i]->size)
+            continue;
+        buf = MPIDIU_get_buf(pools[i]);
+        goto fn_exit;
+    }
+    /* TODO: for now we simply allocate another buffer and return it
+     * after registering the memory. Ideally, when the buffer is
+     * released it should be added to a huge pool so that it can be
+     * reused later on */
+    MPIDIU_buf_t *tmp_buf;
+    MPL_gpu_malloc_host((void **) &tmp_buf, size + sizeof(MPIDIU_buf_t));
+    tmp_buf->next = NULL;
+    tmp_buf->pool = NULL;
+    buf = tmp_buf->data;
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_GPU_GET_BUF);
+    return buf;
+}
+
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_release_buf(void *buf)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_RELEASE_BUF);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_RELEASE_BUF);
+    MPIDIU_buf_t *cont = MPL_container_of(buf, MPIDIU_buf_t, data);
+    if (cont->pool == NULL) {
+        /* Free huge buffer right away */
+        MPL_gpu_free_host(cont);
+        goto fn_exit;
+    }
+    MPIDIU_release_buf(buf);
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_GPU_RELEASE_BUF);
+}
+
+typedef struct MPIDIG_gpu_stage_ctx_t {
+    bool is_send;
+    union {
+        struct {
+            MPI_Aint true_lb;
+        } send;
+
+        struct {
+            void *dst_buf;
+            MPI_Aint count;
+            MPI_Datatype datatype;
+        } recv;
+    } u;
+} MPIDIG_gpu_stage_ctx_t;
+
+/* MPIDIG_gpu_stage_send_buf - allocate stage buffer, copy source data into it, and return pointer */
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_stage_send_buf(const void *buf, MPI_Aint count,
+                                                         MPI_Datatype datatype)
+{
+    char *tmp_buf = (char *) buf;
+    int rc;
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
+
+    tmp_buf = MPIDIG_gpu_get_buf(size);
+    yaksa_request_t request;
+    uintptr_t actual_pack_bytes;
+    rc = yaksa_ipack((const void *) buf, (uintptr_t) size, YAKSA_TYPE__BYTE, (uintptr_t) true_lb,
+                     tmp_buf, (uintptr_t) size, &actual_pack_bytes, &request);
```

Comment:
What are you trying to do here?  Why are you not using yaksa to pack the actual datatype, instead of using `YAKSA_TYPE__BYTE`?  I feel like a lot of this code would simply be deleted if we used yaksa for the actual datatype pack.  In fact, this function would almost entirely be just `yaksa_ipack + wait`, so I don't know if we even need these utility functions.

Reply:
I can reply to this one. You are right we are not packing because we don't have Yaksa integrated yet. So instead we copy the whole buffer for now (including holes in it). Eventually, this will do packing.

Reply:
Oh, yes, I remember you mentioning this now.  Thanks for the reminder.